### PR TITLE
Fix #188; Using distances to filter WERs

### LINKED ISSUE #188: Exclude WER=0 cases from WER reports by using accuracy=0 and not precalculated WERs
The WER calculation should only be done on the final top 10 items within the WER report calculation. This is to prevent high CPU utilization.

### PR DESCRIPTION
**Pull request by tilmankamp**
@kdavis-mozilla This is the slightly optimized WER calculation using ```distance=0``` as ```WER=0``` exclusion criteria. Please take a look.

### CODE REVIEW

File: DeepSpeech.ipynb [hunk 1]
```diff
@@ -1150,73 +1156,88 @@
...
-    "        items.sort(key=lambda a: a[2])\n",
-    "        for a in items:\n",
-    "            print \"-\" * 80\n",
-    "            print \" - WER:    %f\" % a[2]\n",
-    "            print \" - loss:   %f\" % a[3]\n",
-    "            print \" - source: \\\"%s\\\"\" % a[0]\n",
-    "            print \" - result: \\\"%s\\\"\" % a[1] \n",
-    "    print \"#\" * 80\n",
-    "    print"
+    "def get_results_params(data_set):\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: I don't think this function is used any more. I guess this was a leftover from Chris's patch.

**Comment by tilmankamp** [comment]
@Cwiiis @kdavis-mozilla This needs a second review. Sorry.

File: DeepSpeech.ipynb [hunk 2]
```diff
@@ -1145,68 +1151,66 @@
...
    "source": [
-    "Another routine will help collecting partial results for the WER reports. The ```results_tuple``` is composed of an array of the original labels, an array of the corrsponding decodings and an array of the corresponding losses. ```returns``` is built up in a similar way, containing just the unprocessed results of one ```session.run``` call (effectively of one batch). Before splicing them into their corresponding ```results_tuple``` lists, labels and decodings are converted to text. In the case of decodings, for now we just pick the first available path."
+    "Another routine will help collecting partial results for the WER reports. The ```results_tuple``` is composed of an array of the original labels, an array of the corrsponding decodings, an array of the corrsponding distances and an array of the corresponding losses. ```returns``` is built up in a similar way, containing just the unprocessed results of one ```session.run``` call (effectively of one batch). Before splicing them into their corresponding ```results_tuple``` lists, labels and decodings are converted to text. In the case of decodings, for now we just pick the first available path."
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: s/corrsponding/corresponding/

Comment (Cwiiis) [inline DeepSpeech.ipynb]: Would be nice to switch around the sentence that starts with Before, so it reads

> Labels and decodings are converted to text before splicing them into their corresponding `results_tuple` lists.

File: DeepSpeech.ipynb [hunk 3]
```diff
@@ -1274,10 +1255,11 @@
     "def run_inference(session, caption, data_set, results_params=None):\n",
     "    if results_params is None:\n",
     "        # Get the data_set specific graph end-points\n",
-    "        results_params = get_results_params(data_set)\n",
+    "        tower_decodings, tower_labels, _, tower_distances, tower_total_losses, _, _ = get_tower_results(data_set)\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: This hasn't changed in this patch-set, but I worry about recreating this graph every time this function is called, rather than just doing it once. I suppose the flip-side is it can get garbage-collected and associated resources would be freed and the cost of creating the graph over the entire run is miniscule... I have no strong feelings here, as long as you've considered it.

File: DeepSpeech.ipynb [hunk 4]
```diff
@@ -714,10 +714,10 @@
     "    accuracy = tf.reduce_mean(distance)\n",
     "```\n",
     "\n",
-    "Finally, the `total_loss`, `avg_loss`, `accuracy`, the `decoded` batch and the original `batch_y` are returned to the caller\n",
+    "Finally, the `total_loss` and `avg_loss`, the `distance` and `accuracy`, the `decoded` batch and the original `batch_y` are returned to the caller\n",
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: Thanks for the rephrasing.

File: DeepSpeech.ipynb (hunk 3 above)
Comment (tilmankamp) [inline DeepSpeech.ipynb]: To avoid constructing multiple graphs for the same purpose or repeated execution, one can pass in `results_params`.

Comment (Cwiiis) [inline DeepSpeech.ipynb]: Right, but we don't - I would just wonder if this has any impact over a long run where we reconstruct the graph for every validation.

Comment (tilmankamp) [inline DeepSpeech.ipynb]: Good point - for this upcoming checkpointing based scenario, I'll look into graph reuse.

**Comment by tilmankamp** [comment]
@Cwiiis Please take another look.

File: DeepSpeech.ipynb [hunk 5]
```diff
@@ -889,19 +889,21 @@
...
     " * the accuracy of the outcome averaged over the whole batch ```accuracy``` \n",
     " \n",
     "and retain the original ```labels``` (Y).\n",
     " \n",
-    "```decoded```, ```labels```, the optimization gradient and ```total_loss``` are collected into their respective arrays ```tower_decodings, tower_labels, tower_gradients, tower_total_losses``` (dimension 0 being the tower).\n",
+    "```decoded```, ```labels```, the optimization gradient, ```distance```, ```accuracy```, ```total_loss``` and ```avg_loss``` are collected into the corresponding arrays ```tower_decodings, tower_labels, tower_gradients, tower_distances, tower_accuracies, tower_total_losses, tower_avg_losses``` (dimension 0 being the tower).\n",
     "\n",
-    "Finally this new method `get_tower_results()` will return those tower arrays plus (based on ```accuracy```) the averaged accuracy value and (based on ```avg_loss```) the averaged loss value accross all towers."
+    "Finally this new method `get_tower_results()` will return those tower arrays.\n",
+    "In case of ```tower_accuracies``` and ```tower_avg_losses```, it will return the averaged values instead."
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: nit; instead of what? I think you can just remove that word.
